### ISSUE TITLE
add (editable) nonce to confirm screen

### ISSUE THREAD

**Issue by SCBuergel** [MetaMask/metamask-extension#1410]
Currently I can edit the gas limit and gas price on the confirm screen. The nonce is not being displayed. It would be very helpful to at least show the nonce for that transaction but even better would be to make that field editable. That would help in the following situations:
- long pending transactions that I want to "overwrite" with another one that might e.g. have higher gas price
- "oh shit I sent a wrong tx"-situations that might be quickly fixable by "overwriting" it with another transaction
- testing what strategy miners actually use to sort transactions

**Comment by flyswatter** [comment]
I agree we need some method of letting users manage nonces. We may just need to make them more visible across the entire application, from the displays of individual accounts, to transactions.

I've even been thinking maybe clicking a tx should go to a tx detail view (instead of straight to etherscan), and we could show the nonce there, as well as maybe an "Edit" button that allows "Resubmit".

I'd like us to pass this through a little design time before we show it, since it's one field that can completely prevent a transaction from working, but I agree this is important to address.

**Comment by PhABC** [comment]
What is the status on this? The recent spike demand made my 5 GWei transaction completely stuck and I was unaware of the traffic. It's been more than 12h and I would appreciate being able to reset the nonce without going through the hoops of importing my privatekey somewhere else.

**Comment by danfinlay** [comment]
Either this or #2278 is now a very high priority, for the reasons you've described. Hopefully in the next couple weeks.

**Comment by rstormsf** [comment]
please!!!

**Comment by elie222** [comment]
For development this is also a headache. I reset ganache and the nonce is now back to 0, but metamask sees the nonce as being 190 in my browser. I don't want to keep switching accounts.

**Comment by elie222** [comment]
Ah, I see that I can set the nonce in web3:
https://github.com/ethereum/wiki/wiki/JavaScript-API#parameters-25

That helps me for development :)

**Comment by danfinlay** [comment]
For the ganache issues, refer to #2373 and #1999.

**Comment by adrianmcli** [comment]
I see that bounty worthy was added by @danfinlay. How would I go about this if I want to see if I can tackle it?

**Comment by danfinlay** [comment]
Sorry @adrianmcli this one isn't an ideal bounty anymore, because it involves designs that we haven't produced yet, right before we roll out the new UI, so effort on this would be likely to be wasted at the moment.

We should really document our UI/design process flow better as we add more bounties, I realize I keep using "design blocker" as a reason an issue isn't ideal for public contribution yet...

**Comment by dmdque** [comment]
+1 that this is really necessary asap

**Comment by danfinlay** [comment]
@dmdque Could you describe your use case that makes this an asap feature for you?

**Comment by variousred** [comment]
I think _any_ time you need to override a nonce is an ASAP issue ...

**Comment by dmdque** [comment]
If I accidentally sent a bad transaction, I'd be able to cancel it before it gets mined, potentially saving a lot of trouble (and money).

**Comment by danfinlay** [comment]
Ok, that’s fair. We are also going to just add a “cancel” button that sends a higher gas tx with the same nonce, but we could add a setting to show this field (still hiding it from new users).

**Comment by Mecooo** [comment]
Hello,
is there any way to give a command in html/js which will insert a specific nonce?

Like:
......
  web3.eth.sendTransaction({
    to: user_address,
    from: user_address,
    value: web3.toWei('0.001', 'ether'),
    nonce: 274,
.......

Thank you!

**Comment by elie222** [comment]
Yes. That will work.

**Comment by Mecooo** [comment]
Well, it doesn’t work and I don’t even have an idea why :/

**Comment by elie222** [comment]
If you've sent the wrong nonce you should see that in the error message. I
don't think this is a metamask issue, but a web3 one. You should check
through their docs. Do things work when you don't add the nonce field?

Also, most of the time you don't want to be setting the nonce. If you need
to reset the metamask nonce, they have an option for that in the ui.

**Comment by Mecooo** [comment]
The txn itself works.
The aim is to set the same nonce as in one of my pending txns to cause a failure which will abort both txns. As I know this is atm the only way to cancel a pending txn.

In trust wallet it is no problem. There I can set my wished nonce. But in MetaMask isn’t a possibility that I’ve found so far.

**Comment by Mecooo** [comment]
I mean, do you understand what my aim is? Do you guys have an other idea?

**Comment by Mecooo** [comment]
![grafik](https://user-images.githubusercontent.com/42712711/44750081-bfc73800-ab14-11e8-85dd-83cb9410d8b5.png)

With that the txn is starting.. everything is working, but the nonce isn't. It is opening MetaMask with exactly the right txn like I set in the code.. But without the given nonce.

It is ignoring the line with the nonce.. And I have rly no idea why.

I would very appreciate your help! Thank you very much!

**Comment by danfinlay** [comment]
Hi @Mecooo:

No, MetaMask ignores dapp-provided nonces right now. Can you describe the reason you wanted to edit it?

Right now we manage nonces to account, to simplify a variety of possible problems.

**Comment by Mecooo** [comment]
Hi,
there were pending txns in one of my accounts. I sent the ETH to a smartcontract. After a huge gain in gasprices they stucked in the system. After a short while it was not longer worth it to send the ETH to the smart contract. Because of that I wanted to cancel the txn. But there is no way with MetaMask. 

So it worked with trustwallet on my Smartphone. But I want to code something that makes it possible with the PC and MetaMask. I lost a huge amount of money before I had Trustwallet because of this issue, and some others lost theirs as well, cause they couldn't abort without trustwallet. This is the basic reason. I want to give my txn the same nonce as the pending txn that I wanna cancel. The failure would do the rest. As I said.. it worked perfectly on my phone with trustwallet.

**Comment by danfinlay** [comment]
MetaMask uses your currently pending transactions to calculate the correct nonce to give an additional transaction.

When submitting one in place of another pending transaction, the rest of your transactions are still in queue, and so letting users do that has resulted in some dangerous and confusing situations in the past:
https://consensys.zendesk.com/hc/en-us/articles/360004133232-Ghost-Accounts-Ether-Sending-Away-Whenever-Funding-An-Account

To address this problem in a safe way, we are currently implementing the ability to cancel your transactions (https://github.com/MetaMask/Design/issues/25, https://github.com/MetaMask/metamask-extension/issues/4022) in a way that actually keeps users safe from confusing situations like the one above.

One way to work around this is to just ignore the transactions that have been broadcast. You can do this in MetaMask today using the [reset account](https://consensys.zendesk.com/hc/en-us/articles/360004177531-Resetting-an-Account-New-UI-) feature, but you can also just send a transaction from any other client to get the same effect. In this scenario, you send a replacement transaction, and hope for the best. We don't make this easy by default right now, because we've seen it can result in users losing funds in other ways.

Sorry this bit you, but rest assured we are actively working on solving the problem you faced in a way that is even safer than switching between clients.

**Comment by Mecooo** [comment]
> but you can also just send a transaction from any other client to get the same effect. In this scenario, you send a replacement transaction, and hope for the best. 

You mean like TrustWallet, right? Would an other browser with MetaMask work as well?
For example: 
I have these issues in firefox and can just install chrome and insert MetaMask there.. add my private key and launch a txn.. But how without chaning the nonce? Is it because the other client hasn't gotten the new (pending) txn number so far? And why is it possible in Trustwallet. I mean, there you can easily change the nonce. 

Thank you for your answer :)

**Comment by Mecooo** [comment]
Hello again @danfinlay :)
Is it possible to code a external button on my website that will reset MetaMask with its txn list? Probably not, but thought it is worth asking it.

Thank you,
greetings Meco

**Comment by danfinlay** [comment]
> You mean like TrustWallet, right? Would an other browser with MetaMask work as well?

Yes, any other client, including another browser with MetaMask would work in this scenario.

> Is it possible to code a external button on my website that will reset MetaMask with its txn list? 

No.

> I have these issues in firefox and can just install chrome and insert MetaMask there.. add my private key and launch a txn.. But how without chaning the nonce? Is it because the other client hasn't gotten the new (pending) txn number so far? And why is it possible in Trustwallet. I mean, there you can easily change the nonce.

We could very easily let you change the nonce, and we will probably add this under a flag in settings, but by default, we see far more users lose funds and get frustrated by editing their nonces without knowing what they're doing than we get people who want to edit their nonce.

**Comment by Mecooo** [comment]
Well, that is absolutely true! 

Of course there are more people using MetaMask who are not familiar with this option and would probably use it in a wrong way. Provided that it is obvious like changing gas and so on. 

What would you say if you -for the time being- add the possibility to edit it only with some external "tools". So the function is limited to these tools and people have to search for it. So there would be no one who would do it by mistake, because there wouldn't be a direct possibility in the MetaMask interface. And on the other hand you would offer the possibility for those who wanna use it. So those users would do it consciously.. most likely with the knowledge which is important for them.

Win-win in my opinion with a very minimum of risk.

**Comment by Mecooo** [comment]
To be honest.. I thought about this idea as well. But without the cancel feature the user would still have to know how it exactly works and how he has to set the nonce and where to find it on etherscan for example. Sometimes people just think they know it 😉

On my website is everything the have to know and it is very easy to understand.

Isn’t it easy just to let MetaMask take the dapp provided nonce with
...
nonce: 123,
}
Cause you said atm it just ignores it. So just don't ignore it :)

With the cancel feature you are right. After this release there wouldn’t be a reason or benefit for my idea.

**Comment by mehmeta** [comment]
Is anyone working on this? Wouldn't this be helpful in the following scenario? 

1. Send an ETH transaction, with low gas, gets stuck (with nonce a)
2. Realize it takes too long, send a cancellation tx, with a higher gas, also with nonce a.
3. Cancellation tx also takes too long, even though you increased the gas amount. It seems that the "speed up" button only appears on the latest pending tx. So at this point, if I wanted to simply send the first ETH tx with a higher gas amount (rather than canceling), I'm not able to do that (since speed up button only appears on the latest pending tx, which is the cancel tx here). I have to submit a cancellation tx on the 1st ETH tx successfully before I can submit any more transactions.

I'd be happy to work on this feature, if there's any guidance.

**Comment by Mecooo** [comment]
Well I was working on my idea and already finished it. There is just no way as long as MetaMask ignores nonce of external dapps

**Comment by bdresser** [comment]
@mehmeta that's a good point - we should probably persist the "speed up" button on all tx with the lowest nonce. Filed here: https://github.com/MetaMask/metamask-extension/issues/5732

**Comment by bdresser** [comment]
Also relates to https://github.com/MetaMask/metamask-extension/issues/5544

**Comment by ExperimentsWithCode** [comment]
Could really use this feature. Trying to place a bid. My transaction has been stuck for a long time. Tried again with higher gas, also stuck seemingly forever. At this point my bid is useless and i'd rather just submit a new bid over the same nonce but don't have that capability.

**Comment by martinkrung** [comment]
Can you add nonce to edit? This ticket is 2 years old. As it not often used, it would be possible to "hide" it in the settings, like "overwrite nonce for next tx" or so.

**Comment by danfinlay** [comment]
> Can you add nonce to edit? This ticket is 2 years old. As it not often used, it would be possible to "hide" it in the settings, like "overwrite nonce for next tx" or so.

I think adding this as an advanced setting would be the appropriate way to do it. Would be a simple addition, but is probably not going to hit our primary priorities soon, making it a better candidate for a bounty. I would probably want to open a fresh issue specifying that exact feature cleanly, for the sake of easing a bounty hunter's cognitive load. Added as #6757 .

Btw, regarding cancelling transactions above: We have a "cancel" button on pending txs, so nullifying transactions safely is already possible. Try clicking a pending tx to try it!

**Comment by martinkrung** [comment]
Thanks for the new ticket, looking forward. I ended up using an old offline version of myetherwallet: I did a offline tx, where you have to set nonce anyway.

**Comment by doggodfroglog** [comment]
@danfinlay Cancel doesn't work. Metamask lost my transaction, says invalid after speeding up. Can still see old one pending on Etherscan. Need to chance nonce manually now and cancel.

**Comment by doggodfroglog** [comment]
@danfinlay Honestly not even sure how to do it lol. I'm gonna have to import my private keys to offline MEW now because metamask is missing this feature?

**Comment by doggodfroglog** [comment]
![chrome_ouoEuHoOJa](https://user-images.githubusercontent.com/18181181/60604254-4415c780-9d85-11e9-9e7c-7ca396a8b453.png)

**Comment by danfinlay** [comment]
@doggodfroglog help us diagnose that issue, you can download your "state logs" by [following this guide](https://metamask.zendesk.com/hc/en-us/articles/360015290092-How-To-Get-Logs-And-Help-MetaMask-Support-and-Diagnose-Your-Issue), and then sending them to support@metamask.io .

**Comment by fabtho** [comment]
This should fix this, hopefully merged soon: https://github.com/MetaMask/metamask-extension/pull/7089

**Comment by lazaridiscom** [comment]
#7089 looks merged, issue here closable?
